app/static/src/app/components/code/CodeEditor.vue

```diff
@@ -29,23 +29,92 @@ export default defineComponent({
+
+        const getMonacoRange = (line: number, reset = false) => {
+            return {
+                startLineNumber: line,
+                startColumn: reset ? 1 : 0,
+                endLineNumber: line,
+                endColumn: reset ? 1 : Infinity
+            };
+        };
+
+        type EditorStates = "error" | "warning"
+
+        type EditorGlyphs = {
+            [K in EditorStates]: string
+        }
+
+        const editorGlyphs: EditorGlyphs = {
+            error: "fa-solid fa-circle-xmark",
+            warning: "fa-solid fa-triangle-exclamation"
+        };
+
+        const getEditorOptions = (state: EditorStates, message: string) => {
+            return {
+                className: `editor-${state}-background`,
+                glyphMarginClassName: `${editorGlyphs[state]} ${state}-glyph-style ms-1`,
+                hoverMessage: { value: message },
+                glyphMarginHoverMessage: { value: message }
+            };
+        };
+        const getNewDecorations = (lines: number[], state: EditorStates, message: string) => {
+            const errorLineDecorations = lines.map((line) => {
+                return {
+                    range: getMonacoRange(line),
+                    options: getEditorOptions(state, message)
+                };
+            });
+            return errorLineDecorations;
+        };
+        const applyDecorations = (lines: number[], state: EditorStates, message: string, decorations: string[]) => {
+            let decors = decorations;
+            decors = editorInstance.deltaDecorations(decors, getNewDecorations(lines, state, message));
+            oldDecorations.value = decors;
+        };
+        const resetDecorations = (decorations: string[]) => {
+            editorInstance.deltaDecorations(decorations, [{ range: getMonacoRange(1, true), options: {} }]);
+        };
+
+        const updateCode = () => {
+            store.dispatch(`code/${CodeAction.UpdateCode}`, newCode, { root: true })
+                .then(() => {
+                    resetDecorations(oldDecorations.value);
+                })
+                .then(() => {
+                    if (odinModelResponse.value.error) {
+                        const newErrors = odinModelResponse.value.error.line || [];
+                        const message = odinModelResponse.value.error.message || "";
+                        applyDecorations(newErrors, "error", message, oldDecorations.value);
+                    }
+                    if (odinModelResponse.value.metadata?.messages) {
+                        const { messages } = odinModelResponse.value.metadata;
+                        messages.forEach((message) => {
+                            const warningLine = message.line || [];
+                            const warningMessage = message.message || "";
+                            applyDecorations(warningLine, "warning", warningMessage, oldDecorations.value);
```

Comment:
As we discussed, it seems that this could be an issue if there were multiple warnings, since `applyDecorations` calls `deltaDecorations` and sets `oldDecorations` each time, so you would potentially overwrite the first warning with the second? Could collect all decoration info into  single call to `deltaDecoration` (or `createDecorationsCollection`?) instead?

Reply:
agreed, now all the decorations are collected and then applied in one go (note that the odin warnings and odin errors are in different types of payloads so they will never come together)